### ISSUE TITLE
[week3 과제] FoodCollector — Ray 대신 Camera 센서로 학습하기

### ISSUE THREAD

**Issue by linklingj** [pandora-cube/2026-PCUBE-RL-Study#2]
## 목표

week3 실습에서 만든 `FoodCollector`를 **Ray Perception Sensor 없이, Camera Sensor(이미지 관측)만으로** 다시 학습시킵니다.

같은 환경·같은 보상인데 "무엇을 보느냐"만 바꿨을 때 학습이 어떻게 달라지는지(수렴 속도, 최종 보상, 학습 시간) 체감하는 게 핵심입니다.

- Ray: `(태그 원-핫 + 안 맞음 + 거리)` → 이미 정제된 저차원 관측
- Camera: 픽셀 그대로 → CNN 인코더가 "고기 vs 당근"을 **스스로** 구분해야 함

---

## 요구사항

- [ ] Agent에서 `RayPerceptionSensorComponent3D` **제거(또는 비활성화)**
- [ ] `CameraSensorComponent` 부착 + 에이전트 시점 카메라 구성
- [ ] 벡터 관측(속도·frozen 등 보조 관측)은 유지해도 됨 → 대신 **무엇을 남겼는지 명시**
- [ ] 학습 완료 후 `.onnx` 모델을 Behavior Parameters에 물려 **추론 모드 동작 확인**
- [ ] 아래 제출 양식대로 **이 이슈에 댓글**로 결과 공유

### 제출물

| 항목 | 설명 |
|---|---|
| TensorBoard 스크린샷 | `Environment/Cumulative Reward`는 필수. Episode Length, Losses/Value Loss, Policy/Entropy도 같이 있으면 좋음 |
| Inference 영상 | 학습된 `.onnx`로 돌아가는 에이전트 영상 (10~30초, gif/mp4 모두 OK) |
| 총 스텝 수 | 실제로 돌린 step 수 (max_steps가 아니라 **종료 시점 step**) |
| 학습 시간 | 벽시계 기준 (예: `1h 42m`) |
| 평균 보상 | 최종 구간 Cumulative Reward. 가능하면 Ray 버전 수치와 나란히 |
| 학습 설정값 | 카메라 설정 + `configuration.yaml` 주요 하이퍼파라미터 |

---

## 댓글 템플릿 (복사해서 채워주세요)

````markdown
### 결과 요약

| 항목 | 값 |
|---|---|
| run_id | `vis_01` |
| 총 스텝 | 2,000,000 |
| 학습 시간 | 1h 42m |
| 평균 보상 (최종) | 00.0 |
| 평균 보상 (Ray 버전 비교) | 00.0 |
| 실행 환경 | RTX 0000 / CPU 000 |

### 카메라 / 관측 설정

| 항목 | 값 |
|---|---|
| 해상도 (W×H) | 64 × 64 |
| Grayscale | false |
| Observation Stacks | 1 |
| Compression Type | PNG |
| vis_encode_type | simple |
| 남긴 벡터 관측 | 속도(2), frozen/poisoned 플래그 … |
| 음식 색상 | 고기 = ⬛ / 당근 = ⬛ / 벽 = ⬛ |

### 학습 설정 (configuration.yaml 발췌)

```yaml
behaviors:
  FoodCollector:
    trainer_type: ppo
    hyperparameters:
      batch_size: 1024
      buffer_size: 10240
      learning_rate: 0.0005
      beta: 0.005
      epsilon: 0.2
      lambd: 0.95
      num_epoch: 3
    network_settings:
      hidden_units: 256
      num_layers: 2
      vis_encode_type: simple
    max_steps: 2000000
    time_horizon: 64
env_settings:
  num_envs: 1
  num_areas: 1
engine_settings:
  time_scale: 20
```

### TensorBoard

<!-- 스크린샷 첨부 -->

### Inference 영상

<!-- gif / mp4 첨부 -->

### 회고

- 잘된 점 / 막힌 점:
- Ray 대비 체감 차이:
````

---

## 팁

**1. 음식 색을 확실하게 갈라 놓기**
Ray는 태그로 구분하지만 카메라는 **픽셀 색**이 전부입니다. 고기/당근/벽/바닥이 비슷한 색이면 CNN이 끝까지 구분을 못 배웁니다.
- Material 색을 채도 높은 원색으로 (예: 고기 = 빨강, 당근 = 초록, 벽 = 회색, 바닥 = 어두운 회색)
- 저해상도로 줄였을 때도 구분되는지 **사람 눈으로 먼저 확인** — 안 보이면 에이전트도 못 봅니다
- 같은 이유로 **Grayscale은 켜지 마세요**. 색이 곧 정답 신호입니다

**2. 해상도는 과감하게 낮추기**
기본 `84×84`는 FoodCollector엔 과합니다. 이미지 관측은 CNN을 타서 벡터 관측보다 훨씬 느립니다.
- `32×32` ~ `64×64` 권장. Heuristic 모드로 직접 플레이해 보고 "이 정도면 먹을 수 있겠다" 수준으로 맞추기
- `vis_encode_type`별 최소 해상도가 있습니다 — `simple` 20×20 / `nature_cnn` 36×36 / `resnet` 15×15. 너무 낮추면 학습이 시작도 안 되니 주의
- 해상도를 줄인 만큼 스텝을 더 돌릴 수 있습니다

**3. 그 외**
- 카메라는 에이전트 시점으로 두되, **자기 몸이 화면을 가리지 않게** 위치 조정
- 움직임/속도 정보가 필요하면 `Observation Stacks`를 2~3으로 (관측량도 그만큼 늘어남)
- `num_areas`를 늘리면 학습은 빨라지지만 카메라 렌더 비용이 영역 수만큼 붙습니다 — Ray 때보다 보수적으로
- Ray 버전(2M step 기준)만큼 스텝을 돌려도 보상이 낮게 나오는 게 정상입니다. **"안 됐다"가 아니라 "왜 느린가"를 적어주세요**

---

## 참고

- [week3/Sensors.md](https://github.com/pandora-cube/2026-PCUBE-RL-Study/blob/main/week3/Sensors.md) — Camera 센서 설정 항목 설명
- [week3/FoodCollector.md](https://github.com/pandora-cube/2026-PCUBE-RL-Study/blob/main/week3/FoodCollector.md) — 학습 파이프라인 전체
- [week3/TensorBoard.md](https://github.com/pandora-cube/2026-PCUBE-RL-Study/blob/main/week3/TensorBoard.md) — 지표 읽는 법

## 제출 방법

`CONTRIBUTING.md` 규칙대로 개인 과제는 **본인 fork**에 두고, upstream에는 PR하지 않습니다.

```bash
git checkout -b week03-<본인이름>
git push origin week03-<본인이름>
```

이 이슈에 **fork 브랜치 링크 + 위 템플릿 댓글**을 남겨주세요.

**Comment by linklingj** [comment]
# 최재현 과제 제출

### Tensorboard 스크린샷

<img width="906" height="615" alt="Image" src="https://github.com/user-attachments/assets/7915f2fd-3c4d-445c-8a53-04b357982f14" />

### Inference 영상

https://github.com/user-attachments/assets/7d168bd9-0918-4d63-9f46-d2a3da896631

### 기타 학습 정보

step: 2,610,000
학습 시간: 2.05h
평균 보상: 272.8
학습 환경 수: 9

주요 하이퍼 파라미터
- epsilon: 0.2
- lambda: 0.95
- gamma: 0.99
- learning rate: 0.0005
- max steps: 20000000

### 회고
생각보다 ray 센서에 근접한 정확도에 빠르게 도달해서 놀랐다. 더 기회가 있으면 높은 해상도로 실험하면 더 높은 성능을 보일지 실험해보고 싶다.

**Comment by nanen123** [comment]
# 이성원 과제 제출

### Tensorboard
<img width="1241" height="427" alt="Image" src="https://github.com/user-attachments/assets/7d08bbc3-a75d-403b-aab0-24f006540e80" />

### Inference
https://github.com/user-attachments/assets/afd2e81d-289b-466e-87be-c4ed4c7abfb3

### 기타 학습 정보
총 스텝 수 : 1620000 스텝
학습 시간 : 약 2.5시간
평균 보상 : 116.4
학습 환경 수 : 10

주요 하이퍼 파라미터
- epsilon: 0.2
- lambda: 0.95
- gamma: 0.99
- learning rate: 0.0005
- max steps: 20000000

### 회고
처음 학습을 시작했을 때 학습이 제대로 이뤄지지 않았다. (food_run_camera_01,2,3 참고)
원인을 찾으려고 learning rate를 스터디장님 처럼 0.0005로 바꿔보기도 하고, 카메라 위치나 각도를 바꿔보기도 하고, AI에게도 물어봤지만 제대로 된 원인을 찾을 수 없었다. 따라서 설정값이나 agent의 문제는 아니라고 생각했다.
해결 방법으로는 food / badFood 오브젝트의 크기를 1.5배 정도 키우고, 땅에 반 쯤 묻혀서 스폰되는 것을 위로 조금 올렸다.
카메라와 레이 센서의 차이점에서 착안한 문제점이었고, 실제로 이렇게 바꾸니까 학습이 제대로 이뤄졌다.
정말 사소한 문제점이었지만 날린 시간을 생각하면 가슴이 아프다.

**Comment by ksihun** [comment]
# 권시헌 과제 제출

### TensorBoard 스크린샷

<img width="530" height="582" alt="Image" src="https://github.com/user-attachments/assets/2d67ccfa-8cb4-41e0-ad6a-90fce5d8bcb1" />

### Inference 영상

https://github.com/user-attachments/assets/913283d6-cb97-45a2-b4da-336f8545634b

### 기타 학습 정보

- step: 3,000,000
- 학습 시간: 약 2.36h (약 2시간 22분)
- 최종 평균 보상: 12.78
- 최종 Smoothed Cumulative Reward: 15.11
- 학습 환경 수: 9

카메라 설정
- 해상도: 24 × 24
- CNN encoder: simple
- 좋은 음식: 갈색 구체 / 나쁜 음식: 파란 구체

주요 하이퍼파라미터

- epsilon: 0.2
- lambda: 0.95
- gamma: 0.99
- learning rate: 0.0003
- max steps: 3000000
- batch size: 512
- buffer size: 5120

### 브랜치

https://github.com/ksihun/2026-PCUBE-RL-Study/tree/CNN

### 회고

처음에는 6개 TrainingArea와 100만 step으로 학습했으나 학습이 제대로 이루어지지않고, 보상이 후반부스텝까지도 음수로 발생하는 문제가 있었다.
 의심가는 점이 2가지 있었는데
카메라의 위치 설정으로 인해 캐릭터의 갈색 머리카락이 카메라에 잡혀 이를 고기로 인식하는 것이 아닐까 하는 점 , CNN기반이라 학습 속도 자체가 많이 느려서 아직 성능이 충분히 올라오지 않은점 이 두가지가 의심되었다.
그래서 카메라의 위치를 수정하고 학습량을 9개의 TrainingArea와 300만 step학습으로 변경하자 후반 Cumulative Reward가 꾸준히 상승했다.
그래프 상으로 봤을 때 아직 기울기가 완만해지지 않은 것으로 보아 스텝을 좀 더 늘리면 더 높은 성능까지도 끌어올릴 수 있을 것 같았으나, 학습 시간 상 300만 스텝까지만 진행했다. 아마 Learning rate를 좀 더 높혔으면 더 빠른 시간 내에 최고 성능에 도달 할 수 있었을 것 같은 아쉬움이 있었다.

**Comment by linklingj** [comment]
@ksihun 카메라가 1인칭인데다가 fov가 작아서 바로 앞에 있는 구체를 발견하지 못하는게 아닌가 싶네요

**Comment by Cleanhea** [comment]
# 홍민기 과제 제출
### Tensorboard 스크린샷

<img width="576" height="427" alt="Image" src="https://github.com/user-attachments/assets/ce04604c-e18b-41d1-ab6e-d0f6aaba40ad" />

### Inference 영상
https://github.com/user-attachments/assets/d3fcbca6-e7ea-4149-bc6c-7dbbd615790d


기타 학습 정보
step: 3,190,000
학습 시간: 2.327h
평균 보상: 122.19
학습 환경 수: 9

### 주요 하이퍼 파라미터
epsilon: 0.2
lambda: 0.95
gamma: 0.99
learning rate: 0.0005
max steps: 5000000

### Fork Repository
https://github.com/Cleanhea/2026-PCUBE-RL-Study

### 회고
처음 초기 아무것도 건들지 않고 학습을 돌렸는데 전혀 만족스러운 결과가 나오지 않았다.
따라서 여러가지 조치를 시행했는데, 우선 화질 저화에 따른 그림자 구분이 성능에 크게 영향이 갈까 싶긴 했지만 최적화를 위해 Camera View 레이어의 모든 Material을 Unlit으로 바꾸어 그림자를 죽였다.
다음으로 고개를 도리도리잼잼 하는게 맘에 들지 않아서 코드에서 
```
public override void OnActionReceived(ActionBuffers actions)
    {
        m_Forward = Mathf.Clamp(actions.ContinuousActions[0],-1f,1f);
        m_Rotate = Mathf.Clamp(actions.ContinuousActions[1],-1f,1f);

        AddReward(-1f / MaxStep);

        AddReward(-turnPenalty * m_Rotate * m_Rotate);
    }
```
다음과 같이 전 스텝에 비해 고개 돌리는게 많을경우 패널티를 줬는데 이러니까 진정된 것 같아서 기분 좋았다. ^.^

**Comment by linklingj** [comment]
@Cleanhea 그림자 죽인거 좋은 아이디어네요. 턴 패널티 준것도 잘하셨습니다

**Comment by DosangE** [comment]
# 배형준 과제 제출

## Tensorboard 스크린샷

<img width="1280" height="688" alt="Image" src="https://github.com/user-attachments/assets/c3aeb155-efca-4b99-90b8-1bb1550e9546" />

## Inference 영상

https://github.com/user-attachments/assets/dbcfb7a6-b04d-495a-99a0-416c5d87ac1b

## 학습 결과

| 항목 | 결과 |
|---|---:|
| 총 학습 step | 1,620,000 steps |
| 학습 시간 | 약 2.5시간 |
| 최종 평균 보상 | 116.4 |
| 병렬 학습 환경 수 | 10개 |

## 주요 하이퍼파라미터

| 파라미터 | 값 |
|---|---:|
| `epsilon` | 0.2 |
| `lambda` | 0.95 |
| `gamma` | 0.99 |
| `learning_rate` | 0.0005 |
| `max_steps` | 20,000,000 |

## 문제 상황

처음 학습을 시작했을 때는 에이전트의 학습이 제대로 진행되지 않았다.
`food_run_camera_01`, `food_run_camera_02`, `food_run_camera_03` 실행에서 평균 보상이
의미 있게 상승하지 않는 현상을 확인했다.

원인을 찾기 위해 다음을 시도했다.

- 스터디장의 설정을 참고해 학습률을 `0.0005`로 변경했다.
- 카메라의 위치와 각도를 조정했다.
- AI에게 설정과 에이전트 구현의 문제 여부를 검토받았다.

하지만 위 변경만으로는 명확한 원인을 찾거나 학습을 정상화하지 못했다. 이 과정에서
하이퍼파라미터나 에이전트 코드 자체보다는 **환경에서 관측되는 대상의 상태**가 문제일
가능성에 주목하게 되었다.

## 원인 분석과 해결

카메라 센서와 레이 센서의 관측 방식 차이에서 원인을 추론했다. 카메라 기반 관측에서는 음식 오브젝트가 너무 작거나 지면에 묻혀 있으면 에이전트 화면에서 충분히 식별되지 않을 수 있다. 반면 레이 센서는 충돌체를 직접 감지하므로 이런 시각적 문제의 영향을 상대적으로 덜 받는다.

이에 따라 다음과 같이 환경을 수정했다.

1. `food`와 `badFood` 오브젝트의 크기를 약 1.5배로 키웠다.
2. 지면에 반쯤 묻혀 스폰되던 오브젝트의 높이를 조금 올렸다.

수정 후 음식과 나쁜 음식이 카메라 화면에서 더 뚜렷하게 관측되었고, 학습이 정상적으로
진행되기 시작했다. 최종적으로 평균 보상 **116.4**를 달성했다.

## 회고

문제는 설정값이나 알고리즘이 아니라 매우 사소한 환경 구성에 있었다. 특히 카메라 센서를
사용하는 강화학습에서는 에이전트가 실제로 무엇을 보는지 먼저 확인하는 과정이 중요하다.
앞으로는 학습을 장시간 실행하기 전에 카메라 관측 화면에서 목표물의 크기, 위치, 가시성을
우선 점검해야 한다. 작은 환경 문제로 많은 학습 시간을 소모했다는 점은 아쉽지만, 센서별
관측 차이를 실제로 이해하게 된 중요한 경험이었다.
